### DISCUSSION TITLE
Scenario "Bern": Docker web container can't connect to db container

### DISCUSSION THREAD

**Discussion by fduran** [SadServers/sadservers#39] [Scenarios]
Discussions about scenario [Bern](https://github.com/fduran/sadservers/tree/main/scenarios/bern)

**Comment by erjan** [comment]
why is there no canonical solution? i tried to stop and remove wordpress container and run it again using wordpress:sad image.

with this cmd specifying additional env vars:

```
sudo docker run -d --name wordpress \
  -e WORDPRESS_DB_HOST=mysql \    #this is specifying the container name as db host
  -e WORDPRESS_DB_NAME=wordpress \
  -p 80:80 \
  wordpress:sad
```

when running test cmd: `sudo docker exec wordpress mysqladmin -h mysql -u root -ppassword ping`

it gives error:

```
mysqladmin: connect to server at 'mysql' failed
error: 'Unknown MySQL server host 'mysql' (-2)'
Check that mysqld is running on mysql and that the port is 3306.
You can check this by doing 'telnet mysql 3306'
```

i think the solution is valid

**Comment by erjan** [comment]
ok i managed to get it to work, i created my own docker network and restarted both containers with some env vars. 
![Screenshot_85](https://github.com/fduran/sadservers/assets/4441068/a9fc453b-fc21-4a9d-9e17-df02a3fa0c35)


however - should the button "check solution" show green? or is it like jakarta scenario where it has no solution?

or similar to resumable server scenario - where it does not run automated check?

**Comment by fduran** [comment]
Hello. Sometimes scenarios have multiple ways of being solved, this is not necessarily a bad thing since it reflects real situations.
This scenario, unlike Jakarta,  has a solution button and a checker (if the "Check My Solution" button is visible, then it will run the check.sh script). In our case the check is https://github.com/fduran/sadservers/tree/main/scenarios/bern#test and your solution looks correct (pinging - really checking the hostname - the database container from the wordpress container works, it returns "mysqld is alive") . I don't know why in this case the button + check script failed. 

If this situation ever happens again, you can run the ~/agent/check.sh script directly to determine if there's a website issue or a script issue, thanks,

**Reply by akantak** [reply]
@fduran there are multiple issues with that challenge, it looks like a bit unfinished and presenting antipatterns, which is in opposition to the learning properties of SadServers. Let me elaborate here a bit.

1. The most important thing here is mentioning the usage of `mysql` hostname everywhere while using `mariadb` in the container name. The check won't work even if (container name)==(hostname) would work. It is very confusing. Also, this is the issue mentioned by @erjan, because he is also trying to connect to `mariadb` and shows it does work, while check.sh uses `mysql` which does not.

2. The quickest (and very dirty solution) is adding mysql to /etc/hosts in the wordpress container (ex. running `echo "172.17.0.2 mysql" >> /etc/hosts`). But I have to say here that modifying the running docker container is an antipattern, so the check.sh should verify it was not done. The current solution saying "hardcode IP in the configuration" is presenting the same antipattern.

3. As the description mentions using hostnames, my first guess would be to use a user-defined bridge network in docker. So basically shut down existing containers and recreate using a user-defined bridge network in which hostnames connection works (not tested on my own but following ex. https://medium.com/@sydasif78/docker-networking-basics-user-defined-bridge-6eea891ee339). It is a "hard" challenge so it sounds reasonable. But the check would fail anyway due to the 1. issue mentioned.

4. The clue & resolution are out of sync between the README in the GitHub repo vs the web page.

Let me know if you need any help with that, I'm happy to help.

**Comment by fduran** [comment]
@akantak these are very valid points, thanks for the detailed review. I'll review this scenario.
